pandora-server-directory/src/account/accountSecure.ts

```diff
@@ -294,20 +348,101 @@ function GenerateSimpleToken(): string {
 	return randomInt(0, 1000000).toString(10).padStart(6, '0');
 }
 
-const TOKEN_TYPES: Record<AccountTokenReason, 'secure' | 'simple'> = {
-	[AccountTokenReason.ACTIVATION]: 'simple',
-	[AccountTokenReason.PASSWORD_RESET]: 'simple',
-	[AccountTokenReason.LOGIN]: 'secure',
-};
+type TokenType = Record<AccountTokenReason, {
+	generate: () => string;
+	expiration: number;
+	limit: number;
+}>;
+
+const TOKEN_TYPES = {
+	[AccountTokenReason.ACTIVATION]: {
+		generate: GenerateSimpleToken,
+		expiration: ACTIVATION_TOKEN_EXPIRATION,
+		limit: 1,
+	},
+	[AccountTokenReason.PASSWORD_RESET]: {
+		generate: GenerateSimpleToken,
+		expiration: PASSWORD_RESET_TOKEN_EXPIRATION,
+		limit: 1,
+	},
+	[AccountTokenReason.LOGIN]: {
+		generate: () => nanoid(32),
+		expiration: LOGIN_TOKEN_EXPIRATION,
+		limit: 5,
+	},
+} as const satisfies TokenType;
+
+export class AccountToken extends TypedEventEmitter<{
+	tokenDestroyed: AccountToken;
+	extended: AccountToken;
+}> implements Readonly<DatabaseAccountToken> {
+	readonly #reason: AccountTokenReason;
+	#value: string;
+	#expires: number;
+
+	#destroyed = false;
+
+	public get value(): string {
+		return this.#value;
+	}
+	public get expires(): number {
+		return this.#expires;
+	}
+	public get reason(): AccountTokenReason {
+		return this.#reason;
+	}
 
-const TOKEN_EXPIRATION: Record<AccountTokenReason, number> = {
-	[AccountTokenReason.ACTIVATION]: ACTIVATION_TOKEN_EXPIRATION,
-	[AccountTokenReason.PASSWORD_RESET]: PASSWORD_RESET_TOKEN_EXPIRATION,
-	[AccountTokenReason.LOGIN]: LOGIN_TOKEN_EXPIRATION,
-};
+	constructor(token: Readonly<DatabaseAccountToken>) {
+		super();
+		this.#value = token.value;
+		this.#expires = token.expires;
+		this.#reason = token.reason;
+	}
 
-const TOKEN_LIMITS: Record<AccountTokenReason, number> = {
-	[AccountTokenReason.ACTIVATION]: 1,
-	[AccountTokenReason.PASSWORD_RESET]: 1,
-	[AccountTokenReason.LOGIN]: 5,
-};
+	public static create(reason: AccountTokenReason): AccountToken {
+		const { generate, expiration } = TOKEN_TYPES[reason];
+		return new AccountToken({
+			value: generate(),
+			expires: Date.now() + expiration,
+			reason,
+		});
+	}
+
+	public getId(): string {
+		return this.value.substring(0, 16);
+	}
+
+	public validate(): boolean {
+		if (this.#destroyed)
+			return false;
+
+		if (this.expires <= Date.now()) {
+			this.onDestroy();
+			return false;
+		}
+
+		return true;
+	}
+
+	public onDestroy(): void {
+		if (this.#destroyed)
+			return;
+
+		this.#destroyed = true;
+
+		this.emit('tokenDestroyed', this);
+	}
+
+	public extend(): boolean {
+		if (this.#destroyed)
+			return false;
+
+		const { generate, expiration } = TOKEN_TYPES[this.#reason];
+		this.#value = generate();
+		this.#expires = Date.now() + expiration;
```

Comment:
With this code is there really even a reason to have the "extend" functionality? It seems all this does is effectively invalidate old token and create a brand new one that is then returned.

Reply:
this is for security reasons to not keep any token valid over a specified time